nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/io/avro/TestAvroSchemaFactoryService.java

```diff
@@ -0,0 +1,228 @@
+        Schema fileext = attachedFile.schema();
+        assertEquals("fileext", fileext.getName());
+        assertSame(STRING, fileext.getField("name").schema().getType());
+        Schema vignettes = fileext.getField("vignettes").schema();
+        assertEquals("array", vignettes.getType().getName());
+        Schema vignette = vignettes.getElementType();
+        assertSame(STRING, vignette.getField("label").schema().getType());
+        assertSame(LONG, vignette.getField("width").schema().getType());
+        assertSame(LONG, vignette.getField("height").schema().getType());
+        Schema content = vignette.getField("content").schema();
+        assertNull(content.getField("mime-type"));
+        assertSame(STRING, content.getField("name").schema().getType());
+        assertSame(STRING, content.getField("digest").schema().getType());
+        assertSame(STRING, content.getField("encoding").schema().getType());
+        assertSame(STRING, content.getField(context.replaceForbidden("mime-type")).schema().getType());
+        assertSame(LONG, content.getField("length").schema().getType());
+        assertSame(BYTES, content.getField("data").schema().getType());
+    }
+
+    @Test
+    public void testDocumentTypeFactory() {
+        for (DocumentType documentType : schemaManager.getDocumentTypes()) {
+            AvroSchemaFactoryContext context = service.createContext();
+            Schema avro = context.createSchema(documentType);
+            assertNotNull(avro);
+            assertEquals(documentType.getSchemas().size(), avro.getFields().size());
+            // the toString(boolean pretty) call here is to assert no exception is thrown during the call
+            // in some cases the schema can be built but not serialized
+            avro.toString(true);
+        }
+    }
+
+    @Test
+    public void testForbiddenCharsConvertion() {
+        SchemaImpl nuxeo = new SchemaImpl("testSchema", new Namespace("hhtp://test.com/schema/test", "tst"));
+        for (String forbidden : FORBIDDEN) {
+            nuxeo.addField(forbidden, StringType.INSTANCE, null, 0, null);
+            nuxeo.addField(forbidden + FIELD + forbidden, StringType.INSTANCE, null, 0, null);
+            nuxeo.addField(FIELD + forbidden + FIELD, StringType.INSTANCE, null, 0, null);
+        }
+        AvroSchemaFactoryContext context = service.createContext();
+        Schema avro = context.createSchema(nuxeo);
+        assertNotNull(avro);
+        for (String forbidden : FORBIDDEN) {
+            String single = context.replaceForbidden(forbidden);
+            assertEquals(single, avro.getField(single).name());
+            String black = context.replaceForbidden(FIELD + forbidden + FIELD);
+            assertEquals(black, avro.getField(black).name());
+            String white = context.replaceForbidden(forbidden + FIELD + forbidden);
+            assertEquals(white, avro.getField(white).name());
```

Comment:
Could you please also add a set of assertions that show the explicit conversion of forbidden values? Like `assertEquals("dc__colon__title", context.replaceForbidden("dc:title"))` etc.